src/focus-visible.js

```diff
@@ -230,45 +234,57 @@ function init() {
     removeInitialPointerMoveListeners();
   }
 
+  // For some kinds of state, we are interested in changes at the global scope
+  // only. For example, global pointer input, global key presses and global
+  // visibility change should affect the state at every scope:
   document.addEventListener('keydown', onKeyDown, true);
   document.addEventListener('mousedown', onPointerDown, true);
   document.addEventListener('pointerdown', onPointerDown, true);
   document.addEventListener('touchstart', onPointerDown, true);
-  document.addEventListener('focus', onFocus, true);
-  document.addEventListener('blur', onBlur, true);
   document.addEventListener('visibilitychange', onVisibilityChange, true);
-  addInitialPointerMoveListeners();
-
-  document.body.classList.add('js-focus-visible');
-}
 
-/**
- * Subscription when the DOM is ready
- * @param {Function} callback
- */
-function onDOMReady(callback) {
-  var loaded;
+  addInitialPointerMoveListeners();
 
-  /**
-   * Callback wrapper for check loaded state
-   */
-  function load() {
-    if (!loaded) {
-      loaded = true;
+  // For focus and blur, we specifically care about state changes in the local
+  // scope. This is because focus / blur events that originate from within a
+  // shadow root are not re-dispatched from the host element if it was already
+  // the active element in its own scope:
+  scope.addEventListener('focus', onFocus, true);
+  scope.addEventListener('blur', onBlur, true);
 
-      callback();
-    }
-  }
+  if (scope.nodeType === Node.DOCUMENT_NODE) {
+    document.documentElement.classList.add('js-focus-visible');
 
-  if (['interactive', 'complete'].indexOf(document.readyState) >= 0) {
-    callback();
-  } else {
-    loaded = false;
-    document.addEventListener('DOMContentLoaded', load, false);
-    window.addEventListener('load', load, false);
+    // We detect that a node is a ShadowRoot by ensuring that it is a
```

Comment:
Putting this comment inside the previous if block is breaking my brain a little bit, though I can see why you did. Maybe switch the order of these if/else statements? It would be _marginally_ less efficient for the "normal" case, I guess, but more readable.

Or, perhaps a switch statement over `scope.nodeType` ?

Reply:
Thanks @alice I'll try to rework this so that it is less of an eyesore

Reply:
I switched the order of the conditions around. The comments are hopefully more cleanly formatted now.